### PR TITLE
Start setting up the improved W&B integration

### PR DESCRIPTION
**Pull request by AyushExel**
This PR adds 2 helper files that are at the center of the new and improved integration.

@glenn-jocher these are new utility files and won't cause and merge conflicts. Once, we have these files here, we can then start utilizing them for improving the existing integration one step at a time. Right now, these files will remain unused. Due to merge conflicts in my branch, this workflow seemed to be the best. This way, we can work on the integration step-by-step instead of merging all features at once. 

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhanced W&B Integration and Dataset Management in YOLOv5.

### 📊 Key Changes
- Added `self.path` attribute to `LoadImagesAndLabels` for better path handling.
- Added two new files, `log_dataset.py` and `wandb_utils.py`, to improve logging and dataset handling with Weights & Biases (W&B).
- `log_dataset.py` allows the creation of W&B dataset artifacts and updates `data.yaml` with artifact links.
- `wandb_utils.py` introduces `WandbLogger` with methods to set up training, log datasets and models, and download dataset artifacts.

### 🎯 Purpose & Impact
- 🔄 Facilitates data versioning and reproducibility of machine learning experiments using W&B artifacts.
- ⚙️ Enhances dataset management by enabling automatic dataset downloading and labeling directly within code.
- 📈 Allows for better tracking and visualization of the training progress and results on W&B.
- 🛠 Users can potentially expect more robust model training pipelines, easier sharing of datasets/models, and streamlined experiment tracking for YOLOv5.

### CODE REVIEW

**Comment by AyushExel** [comment]
I have no idea why one test failed. This PR doesn't change anything in the existing codebase. @glenn-jocher can you please have a look?

Screenshot (323)

**Comment by glenn-jocher** [comment]
> I have no idea why one test failed. This PR doesn't change anything in the existing codebase. @glenn-jocher can you please have a look?
> 
> Screenshot (323)

Maybe a few percent of the time a check will fail for unrelated reasons unfortunately. It would be nice to be automate reruns of failed CI tests in these situations. Currently I have to manually trigger a CI rerun to clear it up, which is a bit of a waste of time, but I don't have the actions yaml expertise to code up my rerun on fail idea myself. Do you know anyone who might be able to do this?

I'll rerun the check manually now, should be fine.

**Comment by glenn-jocher** [comment]
@AyushExel the top level directory is reserved for the 3 main python files (train, test, detect), and we'd like to keep it as clutter free as we can, especially to help newcomers who are browsing the code for the first time understand the structure more easily. Would you be able to combine the added functionality into utils/wandb_utils.py? If not, or if the follow on updates will require additional files, the best implementation would be to create a utils/wandb directory to hold these files, similar to what we've done for utils/google_app_engine.

**Comment by AyushExel** [comment]
> > I have no idea why one test failed. This PR doesn't change anything in the existing codebase. @glenn-jocher can you please have a look?
> > Screenshot (323)
> 
> Maybe a few percent of the time a check will fail for unrelated reasons unfortunately. It would be nice to be automate reruns of failed CI tests in these situations. Currently I have to manually trigger a CI rerun to clear it up, which is a bit of a waste of time, but I don't have the actions yaml expertise to code up my rerun on fail idea myself. Do you know anyone who might be able to do this?
> 
> I'll rerun the check manually now, should be fine.

Yes I know someone who might be able to help. Asking him to take a look.

**Comment by AyushExel** [comment]
> @AyushExel the top level directory is reserved for the 3 main python files (train, test, detect), and we'd like to keep it as clutter free as we can, especially to help newcomers who are browsing the code for the first time understand the structure more easily. Would you be able to combine the added functionality into utils/wandb_utils.py? If not, or if the follow on updates will require additional files, the best implementation would be to create a utils/wandb directory to hold these files, similar to what we've done for utils/google_app_engine.

@glenn-jocher  Totally agreed. Changed the location of both files to `utils/wandb/*` 
Hopefully, I'll be able to fix the problem of steps not being in sync with epochs tomorrow using this setup.

**Comment by glenn-jocher** [comment]
@AyushExel great, thanks! Changes look good. I recently updated a missing wandb package message, I will update your branch with this new message, and an import for the colorstr() dependency (to give it a fun color now).

**Comment by glenn-jocher** [comment]
@AyushExel I've updated the PR a bit to align it more closely with current master. The dataloader calls in log_datasets.py require additional changes though I think, as they've diverged from master. See train.py here:
https://github.com/ultralytics/yolov5/blob/b5d851d653a974e48c47b9fac7699360747c3520/train.py#L181-L197

File: utils/datasets.py [hunk 1]
```diff
@@ -345,7 +345,8 @@ def __init__(self, path, img_size=640, batch_size=16, augment=False, hyp=None, r
         self.mosaic = self.augment and not self.rect  # load 4 images at a time into a mosaic (only during training)
         self.mosaic_border = [-img_size // 2, -img_size // 2]
         self.stride = stride
-
+        self.path = path
```
Comment (AyushExel) [inline utils/datasets.py:351]: I have added this additional change to be able to access that dataset path using `dataloader.dataset.path`. This is also a piece of input info that wasn't really being saved as class variable.

**Review by AyushExel — COMMENTED** [review]
@glenn-jocher I have re-organized the `log_dataset.py` a bit. I have added some arguments from the updated train.py but I've improvised a little bit. 
* I've fixed the ``stride` to 32 as getting stride dynamically requires to load the model, and this script is meant only to log the dataset.
* I'm not using `augment=True` for trainloader because it makes the `shapes=NoneType`. Also, augmentation is not required for logging.
*  I've also set `rank` to -1 because I guess DDP isn't really useful for logging. We're just making 1 pass through all images.

I've added a very minor change in datasets.py which proves useful in logging, mentioned below. 
Once this is done, I'll make a PR addressing steps Vs epochs problem.

File: utils/datasets.py (hunk 1 above)
Comment (glenn-jocher) [reply utils/datasets.py:351]: Sure looks good!

**Comment by glenn-jocher** [comment]
@AyushExel everything sounds good, except the stride may be an issue. Max stride for default models (with P3-P5 outputs) is 32 pixels (P3 is 8 pixels for small objects, P4 is 16 for medium objects, then P5 is 32 for large objects), but larger models with P6 (64 pixels) like https://github.com/ultralytics/yolov5/blob/master/models/hub/yolov5-p6.yaml and P7 (128 pixels) will have larger strides. The stride is important because it informs the dataloader of the padding constraints. 

For example if --img is 640 and a 600x500 size image is loaded, the image stretches (with constant aspect ratio) so that the long side reaches 640 pixels, with the short side padded up to the next stride-multiple, i.e. 640/600*500 = 640x533 stretched image, nearest ceiling multiple of stride 32 is 544, so 11 pixels of padding are required to be added to the short side. The P6 model inference will break in this case, as 544 is not a multiple of 64.

**Comment by glenn-jocher** [comment]
@AyushExel let's do a meeting to figure this out, maybe there's a way to simply things or bypass the stride requirement. Can you select a time here?
https://calendly.com/ultralytics/meeting

File: utils/wandb_logging/wandb_utils.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+        if path.startswith(WANDB_ARTIFACT_PREFIX):
+            dataset_artifact = wandb.use_artifact(remove_prefix(path,WANDB_ARTIFACT_PREFIX)+":"+alias)
+            if dataset_artifact is None:
+                logger.error('Error: W&B dataset artifact doesn\'t exist')
+                raise ValueError('Artifact doesn\'t exist')
+            datadir = dataset_artifact.download()
+            labels_zip = datadir+"/data/labels.zip"
+            shutil.unpack_archive(labels_zip, datadir+'/data/labels', 'zip')
+            print("Downloaded dataset to : ", datadir)
+            return datadir, dataset_artifact
```
Comment (AyushExel) [inline utils/wandb_logging/wandb_utils.py]: Here the images and labels are downloaded as artifacts. And these files will then be loaded by train.py itself in traditional fashion, using the correct stride. Basically, this `datadir` is the original dataset that was uploaded as an artifact, and this dir can be used to load it again, with the correct stride.

**Review by AyushExel — COMMENTED** [review]
@glenn-jocher  I have set up a meeting. Although, I think the stride won't matter in case of logging datasets because it's logging the original images using their paths. Later these images will be downloaded and loaded using the data loader in `train.py` with the correct stride. So, the stride won't really matter.

**Comment by glenn-jocher** [comment]
@AyushExel I've made a few updates here:

- Replaced FullLoader with SafeLoader (https://github.com/ultralytics/yolov5/pull/1962)
- PEP8 reformat on both files to bring them into conformity with the rest of the repo
- Remove dataset checks as redundant with train.py

I noticed a few additional areas for attention:
- Directory and file concatenation should be updated across PR to use Path() objects from pathlib for robustness across operating systems (replacing instances of str + str, and wrapping hardcoded paths, i.e. Path('hardcoded/path')).
- check_dataset(data_dict) will autodownload the dataset by itself if not found locally as I mentioned over the phone. This functionality is already in master and utilizes the download: field in your data.yaml. This is present in your PR on L23 of wandb_logging.
- The device selection in the PR need to conform to the user requested device, i.e. if --device cpu is passed for training, your device selection line will not currently respect that.
- Can we rename wandb_logging/ to wandb/ ?
- The dataloader arguments can largely all be eliminated if all we need is the image paths. See my autoanchor implementation for an example of a minimal dataloader call (EDIT: dataloader seems unnecessary, _dataset_ should suffice as in autoanchor example):
https://github.com/ultralytics/yolov5/blob/d9212140b355b84e85a473be590720eb8221766c/utils/autoanchor.py#L103

I haven't reviewed the second file yet, but can you please address the above points and let me know, and then I will review again? Thanks!

**Comment by AyushExel** [comment]
@glenn-jocher I addressed these points:
*  Replaced dataloader with LoadImagesAndLabels, eliminating most of the dataloader arguments 
*  Using `Path() / 'name'` instead of str + str (WANDB_PREFIX is still str though)
* Not using check_dataset
* Eliminated --device, as we're just logging stuff and not using any GPU intensive tasks

Note:
* The import order here might seem a bit out of place, but importing `WandbLogger` ensures that `utils` is added to the system path which eliminates the need for relative imports.
```
from wandb_utils import WandbLogger
from utils.datasets import create_dataloader, LoadImagesAndLabels
```
* haven't replaced 'wandb_logging' folder name 'wandb' as it causes problems while importing wandb.
* As I do most of this development on GCP VM, I can't check for PEP8 compatibility.

Let me know if I missed something

**Comment by glenn-jocher** [comment]
> @glenn-jocher I addressed these points:
> 
> * Replaced dataloader with LoadImagesAndLabels, eliminating most of the dataloader arguments
> * Using `Path() / 'name'` instead of str + str (WANDB_PREFIX is still str though)
> * Not using check_dataset
> * Eliminated --device, as we're just logging stuff and not using any GPU intensive tasks
> 
> Note:
> 
> * The import order here might seem a bit out of place, but importing `WandbLogger` ensures that `utils` is added to the system path which eliminates the need for relative imports.
> 
> ```
> from wandb_utils import WandbLogger
> from utils.datasets import create_dataloader, LoadImagesAndLabels
> ```
> 
> * haven't replaced 'wandb_logging' folder name 'wandb' as it causes problems while importing wandb.
> * As I do most of this development on GCP VM, I can't check for PEP8 compatibility.
> 
> Let me know if I missed something

Thanks, will review again today!

**Comment by glenn-jocher** [comment]
@AyushExel ok I've reviewed and updated a bit, and commented some areas that seemed to benefit from updates but wanted you to review first. 

One important note is that we have a `python train.py --nosave` options, which only saves last.pt (and copies it as best.pt) on final epoch. Will the new code play well with this?

**Comment by AyushExel** [comment]
@glenn-jocher Yes, we can make sure to use `--nosave` as it's intended. These changes won't affect `train.py` until we start using the `WandbLogger` in the main training script. Those are smaller changes. Let's take that up in the next PR.
Also, I'm out this week, so I'll probably be slow to respond.

**Comment by AyushExel** [comment]
Hey @glenn-jocher if there are no hiccups in this PR, let's get it merged. We can then setup the new Integration next week using these files.

**Comment by glenn-jocher** [comment]
@AyushExel will do! Have a good week off.

**Comment by glenn-jocher** [comment]
@AyushExel sorry, I noticed there's no responses on the 7 review items I opened above. Please review and respond when you have time next week, thanks!

**Review by AyushExel — COMMENTED** [review]
@glenn-jocher I have marked all files as viewed. I don't see any specific review request here.

File: utils/wandb_logging/log_dataset.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
+import argparse
+from pathlib import Path
+
+import yaml
+
+from wandb_utils import WandbLogger
+from utils.datasets import LoadImagesAndLabels
+
```
Comment (AyushExel) [inline utils/wandb_logging/log_dataset.py:8]: @glenn-jocher All the changes that you made look good. I have changed the import order here because importing `WandbLogger` makes sure that the utils directory is added to the system path, eliminating the need for relative imports.

**Comment by AyushExel** [comment]
@glenn-jocher this looks good to me now. Let me know if there is something else that you want me to review :)

**Comment by AyushExel** [comment]
@glenn-jocher hey I'm back from vacation. Let me know if there's anything to be done here. I'm waiting for this to get merged, and then we can make a quick update PR to refine the integration and introduce cool new features.

**Comment by glenn-jocher** [comment]
@AyushExel hey buddy. So I annotated a bunch of items a few days ago. Maybe I did it incorrectly and you can't see it, I'm still a bit of a novice on the PR review process. I can implement most of these myself, but some of them I wanted to run them by you. Here's the items:
Screen Shot 2021-01-29 at 2 19 00 PM

**Comment by AyushExel** [comment]
@glenn-jocher  I cannot see these awesome reviews on my end. Anyway thanks for pointing these out. I've addressed the following.
* **L119 is redundant. Removed it**
* **L102 - L109 seems redundant. Removed it** I was actually planning something around this in the lines of "deployment".  I thought of having a separate artifact called "final_models" with only the final models for each experiment for easier deployment of yolov5 models directly from the terminal. Right now it seems a bit far fetched. But let me know if you have any plans to support deployment in the future. I've gotten rid of it for now.
* **Using single line assert statements replacing the 3-line value error checks**
* **L120- matrix values are assigned to a tuple. This works fine** I'm extracting the matrix values in the code later on.
* **redundant `import os`. Removed it** . Now using the relative imports as you suggested.

Let me know if I missed something

**Comment by glenn-jocher** [comment]
@AyushExel thanks! /rebase

**Comment by AyushExel** [comment]
@glenn-jocher I have rebased this branch and now it's up-to-date with the master branch. Let's get this merged :)

**Comment by glenn-jocher** [comment]
@AyushExel ok looks good now, thanks for the changes!
